pkg/distributor/distributor.go

```diff
@@ -769,7 +769,11 @@ func (d *Distributor) prePushHaDedupeMiddleware(next PushFunc) PushFunc {
 	}
 }
 
-func (d *Distributor) prePushRelabelMiddleware(next PushFunc) PushFunc {
+// prePushRelabelAndSortMiddleware is responsible for applying relabel rules as
+// well as sorting labels. The sort step is critical because ingesters assume
+// that labels are sorted in the push request. Without this, ingesters will
+// write a corrupted block index.
```

Comment:
We could also add check into TSDB writing code to avoid writing corrupted indexes.

Reply:
(I'm suggesting that, as my [additional check in TSDB writing code](https://github.com/prometheus/prometheus/pull/8085) just got merged after 3+ years 😄)